stacks-signer/src/coordinator.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright (C) 2013-2020 Blockstack PBC, a public benefit corporation
+// Copyright (C) 2020-2024 Stacks Open Internet Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::time::Instant;
+
+use blockstack_lib::chainstate::burn::ConsensusHashExtensions;
+use slog::slog_debug;
+use stacks_common::debug;
+use stacks_common::types::chainstate::ConsensusHash;
+use stacks_common::util::hash::Sha256Sum;
+use wsts::curve::ecdsa;
+use wsts::state_machine::PublicKeys;
+
+/// TODO: test this value and adjust as necessary. Maybe make configurable?
+pub const COORDINATOR_OPERATION_TIMEOUT_SECS: u64 = 300;
+
+/// TODO: test this value and adjust as necessary. Maybe make configurable?
+pub const COORDINATOR_TENURE_TIMEOUT_SECS: u64 = 600;
+
+/// The coordinator selector
+#[derive(Clone, Debug)]
+pub struct CoordinatorSelector {
+    /// The ordered list of potential coordinators for a specific consensus hash
+    coordinator_ids: Vec<u32>,
+    /// The current coordinator id
+    coordinator_id: u32,
+    /// The current coordinator index into the coordinator ids list
+    coordinator_index: usize,
+    /// The last message received time for the current coordinator
+    pub last_message_time: Option<Instant>,
+    /// The time the coordinator started its tenure
+    tenure_start: Instant,
+    /// The public keys of the coordinators
+    public_keys: PublicKeys,
+}
+
+impl From<PublicKeys> for CoordinatorSelector {
+    /// Create a new Coordinator selector from the given list of public keys
+    fn from(public_keys: PublicKeys) -> Self {
+        let coordinator_ids =
+            CoordinatorSelector::calculate_coordinator_ids(&public_keys, &ConsensusHash::empty());
```

Comment:
Don't we always want to be taking the current burnchain tip consensus hash?

Reply:
This was also a suggestion I made. Coordinator selection requires some more work -- as currently implemented, the issue is that a change in the burnchain triggers reselection of coordinators, which really shouldn't happen. Coordinators only need to be changed if there's a problem with the current coordinator -- a bitcoin block arriving doesn't indicate that, and was causing the signer set to continuously restart DKG in test setups. My suggestion was to just use a fixed ConsensusHash for now to avoid that issue. Coordinator selection can and should be dealt with in a separate PR.

Reply:
Sure, that makes sense.  That also addresses a concern of mine about the change in coordinator having some degree of clock drift to it (e.g. there'd be bouts of time where some nodes think a stale coordinator is still the current coordinator).